lib/src/license_detection/license_detector.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/third_party/diff_match_patch/diff.dart';
+import 'package:source_span/source_span.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+part 'confidence.dart';
+part 'crc32.dart';
+part 'license.dart';
+part 'primary_filter.dart';
+part 'token_matcher.dart';
+part 'tokenizer.dart';
+
+// Load corpus licenses.
+final _licenses = loadLicensesFromDirectories(_directories);
+
+// WIP: Returns a list of detected licenses whose
+// confidence score is above a certain threshold.
+List<LicenseMatch> detectLicense(String text, double threshold) {
```

Comment:
Also shouldn't we do something to avoid overlapping licenses..

Say we have a unknown text as follows:
```
aa aa aa bb bb cc cc
```

We find the following matches:
 1. `aa aa aa` to be license A (confidence 0.9), 
 2. `bb bb` to be license B (confidence 0.9), 
 3. `cc cc` to be license C (confidence 0.9), 
 4. `aa bb` to be license D (confidence 0.8).

Can't we use confidence and offset in the unknown text to see that result (4) is unlikely / impossible.
Because it's conflicting with result (1) and (2) which have higher confidence.

So detecting that two results are overlapping and then preferring the one with the highest confidence, throwing the other away because it's impossible.


---------------------------------------

This should also help with cases where we corpus license is a subset of another corpus license. As is the case when we have the "END OF TERMS" in licenses like GNU GPL.

Reply:
Maybe we should do something along the lines of:
```
class DetectionResult {
  List<LicenseMatch> matches;
  int unknownCharacters;
}

DetectionResult detectLicense(String text, double threshold) {
...
```

So that the caller can know how many chars were unknown in the `LICENSE` file..

Reply:
This is done,